pkg/collector/python/check.go

```diff
@@ -330,6 +345,80 @@ func (c *PythonCheck) ID() check.ID {
 	return c.id
 }
 
+// GetDiagnoses returns the diagnoses cached in last run or diagnose explicitly
+func (c *PythonCheck) GetDiagnoses() ([]diagnosis.Diagnosis, error) {
+	// Lock the GIL and release it at the end of the run (will crash otherwise)
+	gstate, err := newStickyLock()
+	if err != nil {
+		log.Warnf("failed to cancel check %s: %s", c.id, err)
+		return nil, err
+	}
+	defer gstate.unlock()
+
+	// Get serialized diagnoses. CGO packaging is implemented in
+	//Three|Two::getCheckDiagnoses(RtLoaderPyObject* check) methods
+	// in datadog-agent\\rtloader\\three\\three.cpp (and two.cpp).
+	//
+	// It is serialized as JSON by Python. Handcrafted and significantly more complicated
+	// manual serialization was only 2-2.5 times faster and hence not worth it for low-rate calls like this
+	pyDiagnoses := C.get_check_diagnoses(rtloader, c.instance)
+	if pyDiagnoses == nil {
+		return nil, nil
+	}
+	defer C.rtloader_free(rtloader, unsafe.Pointer(pyDiagnoses))
+
+	// Deserialize it
+	s := C.GoString(pyDiagnoses)
+	var diagnosesJSON []diagnosisJSON
+	err = json.Unmarshal([]byte(s), &diagnosesJSON)
+	if err != nil || diagnosesJSON == nil {
+		return nil, err
+	}
+
+	// Copy it over (in future we may be able to deserialize
+	// directly into target data)
+	var diagnoses []diagnosis.Diagnosis
```

Comment:
Good idea. I have tried diagnoses := make([]diagnosis.Diagnosis, len(diagnosesJSON)) which forced me to developed indexed driven enumeration and access which is probably even less performant. I am still learning Go :(. Thank you.

Reply:
FWIW, assigning by index is a frequent pattern, and I don't think there would be any perf penalty to doing so.

For example, and there are many more throughout the code:
https://github.com/datadog/datadog-agent/blob/7b6d467eebaf7ad716d7cbc61f3a4cd43824de41/pkg/config/environment_containers.go#L144-L147

Reply:
Would be interesting to see. Hopefully optimizer would eliminate indirection in memory access since I would have to do it as many times as fields number. But it also look worse code wise.